### ISSUE TITLE
CFE doesn't allow void to void assignment

### ISSUE THREAD

**Issue by MichaelRFairhurst** [dart-lang/sdk#33934]
The following should have no errors:

```
void f(void x) {}                                                                
void voidFn() {}                                                                 
                                                                                 
void main() {                                                                    
  f(voidFn());                                                                   
}
```

This is critical for mockito, which uses a template argument to accomplish the same thing:

```
void f<T>(T x) {}                                                                
void voidFn() {}                                                                 
                                                                                 
void main() {                                                                    
  f(voidFn());                                                                   
}   
```

Both of these should work. @peter-ahe-google

**Comment by srawlins** [comment]
When was this introduced. dev.68.0 is happy. I haven't tried later than that.

**Comment by MichaelRFairhurst** [comment]
@pq is running into it in this case:

```
void x() {
  return print('foo'); // this is legal
}
```

As I understand it, this is complicating the latest fix of package:analyzer

**Comment by MichaelRFairhurst** [comment]
I don't think this ever got documented anywhere. I remember thinking I should update the language tests to test this stuff, but did not because it can be so much work to just mark all those new tests as failing.

**Comment by pq** [comment]
Thanks @MichaelRFairhurst!

Actually, this time it's not analyzer but related to `coveralls` which is breaking on a linter PR:

https://travis-ci.org/dart-lang/linter/jobs/406425911

```
Failed to precompile dart_coveralls:dart_coveralls:
file:///home/travis/.pub-cache/hosted/pub.dartlang.org/dart_coveralls-0.6.0+1/bin/src/report.dart:16:29: Error: This expression has type 'void' and can't be used.
    if (res["help"]) return print(parser.usage);
```

Still investigating.

**Comment by MichaelRFairhurst** [comment]
> Actually, this time it's not analyzer but related ... breaking on a linter PR

Well that's a big improvement!

**Comment by pq** [comment]
😄 

Some more context:

https://github.com/pq/dart-coveralls/blob/de6c86184e247c92b89dd314ec00e98b30b52f07/bin/src/report.dart#L22

TL;DR coveralls is returning void in an async context which *should* be OK?

**Comment by pq** [comment]
More specifically, the case is:

```dart
Future foo() async {
  return print('whoops');
}

```

**Comment by MichaelRFairhurst** [comment]
repro for the coveralls issue:

```
import 'dart:async';                                                             
                                                                                 
Future voidFn() async {                                                          
  return print('foo');                                                           
}
```

Note that this works fine:

```
dynamic voidFn() {                                                          
  return print('foo');                                                           
}
```

**Comment by MichaelRFairhurst** [comment]
This also should not have been an error:

pkg/analyzer/lib/src/dart/analysis/library_context.dart : 98

```
      logger.run('Append library files', () {                                    
        return appendLibraryFiles(targetLibrary);                                       
      }); 
```

I don't think this should have been an error; the type of `run` should have been inferred as `() -> dynamic`, and then `appendLibraryFiles` should have been OK to return, as void values are OK to return within a dynamic context.

I'm less sure on this one though. May have been an analyzer bug? In any case this would be a breaking change compared to the previous behavior so we'd want to be very sure it needed to change if we want to change it. @eernstg @lrhn

**Comment by eernstg** [comment]
@MichaelRFairhurst said:

> I don't think this ever got documented anywhere.

Just a drive-by comment, if I understood correctly what 'this' refers to here, it is actually covered in the [feature spec](https://github.com/dart-lang/sdk/blob/master/docs/language/informal/generalized-void.md). We have the following rule:

> An actual parameter expression corresponding to a formal
> parameter whose statically known type annotation is the type
> void may have the type void.

and there is a similar rule for `void v = voidExpression;`. Those rules were added in order to allow for "assigning void to void" in several situations. For `return e`, we have had such a rule for a long time, because it was already a consequence of the type rules in Dart 1.

**Comment by peter-ahe-google** [comment]
> This also should not have been an error:
> 
> pkg/analyzer/lib/src/dart/analysis/library_context.dart : 98
> 
>       logger.run('Append library files', () {                                    
>         return appendLibraryFiles(targetLibrary);                                       
>       }); 
> I don't think this should have been an error; the type of run should have been inferred as () -> dynamic, and then appendLibraryFiles should have been OK to return, as void values are OK to return within a dynamic context.
> 
> It was fixed in the analyzer source, and the change is certainly a fair change, but this may affect other packages.
> 
> I'm less sure on this one though. May have been an analyzer bug? In any case this would be a breaking change compared to the previous behavior so we'd want to be very sure it needed to change if we want to change it. @eernstg @lrhn

I agree. When I read the code, I misunderstood it and didn't think more about it.

**Comment by peter-ahe-google** [comment]
> More specifically, the case is:
> 
>     Future foo() async {
>       return print('whoops');
>     }

AFAICT, this isn't a valid return see [invalid_returns.md](https://dart-review.googlesource.com/c/sdk/+/60401/6/docs/language/informal/invalid_returns.md).

**Comment by MichaelRFairhurst** [comment]
> > More specifically, the case is:
> > 
> >     Future foo() async {
> >       return print('whoops');
> >     }
>
> AFAICT, this isn't a valid return see [invalid_returns.md](https://dart-review.googlesource.com/c/sdk/+/60401/6/docs/language/informal/invalid_returns.md).

It seems you're right, though this also seems inconsistent (to me). @eernstg thoughts?

```
// valid
dynamic f() { return print(); }
// invalid
Future<dynamic> fPrime() async { return print(); }
```

I suppose this may have been handled differently because the former definition can "look" void (`f(){}`) where the latter cannot "look" void (`Future f(){}`)?

**Comment by leafpetersen** [comment]
I don't understand why you say that 

```dart
Future foo() async {
  return print('whoops');
}
```

is not allowed by https://dart-review.googlesource.com/c/sdk/+/60401/6/docs/language/informal/invalid_returns.md.

On line https://dart-review.googlesource.com/c/sdk/+/60401/6/docs/language/informal/invalid_returns.md#56 , there is the following:

* `return exp;` where `exp` has static type `S` is a valid return if:
  * `flatten(T)` is `dynamic` or `Null`

where `T` is the return type of the function.  `flatten(Future)` is `dynamic`, so this is a valid return.  Am I missing something?

**Comment by MichaelRFairhurst** [comment]
Thanks @leafpetersen, I missed that part! OK, glad to see that its consistent, I think that is the right approach. And it matches current analyzer behavior so that's one fewer deviation to consider.

**Comment by devoncarew** [comment]
From discussion on https://github.com/dart-lang/sdk/issues/34031, this issue is causing spurious warnings when running the Flutter tests and in the Fuchsia build; /cc @Hixie @tvolkert. We'll want to come up with a solution here - either fixing this issue, or reverting the warnings CL.

**Comment by kmillikin** [comment]
I'll take this one.
